middleware/proxy.go

```diff
@@ -359,6 +359,10 @@ func ProxyWithConfig(config ProxyConfig) echo.MiddlewareFunc {
 					c.Set("_error", nil)
 				}
 
+				// Reuse context from echo request
+				// This is needed to access context values in modifyResponse.
+				req = req.WithContext(c.Request().Context())
```

Comment:
I do not understand what this line does. `req` variable is created at line 321 `req := c.Request()`  I do not see any point there that request is replaced with somethings else or there is that method where that `req` context is being changed?

Please provide testcase or small example that I can try with debugger

Reply:
@x1h0 and @TomKrouss  maybe we should make it possible to configure https://github.com/golang/go/blob/d72f4542fea6c2724a253a8322bc8aeed637021e/src/net/http/httputil/reverseproxy.go#L125  which you can use modify request before proxying is done.

at the moment I do not undestand that situation where this change would help you.

Reply:
ok, I think I see your point.

```go
package main

import (
	"context"
	"github.com/labstack/echo/v4"
	"github.com/labstack/echo/v4/middleware"
	"github.com/stretchr/testify/assert"
	"net/http"
	"net/http/httptest"
	"net/url"
	"testing"
)

type customBalancer struct {
	target *middleware.ProxyTarget
}

func (b *customBalancer) AddTarget(target *middleware.ProxyTarget) bool {
	panic("not implemented")
	return true
}

func (b *customBalancer) RemoveTarget(name string) bool {
	panic("not implemented")
	return false
}

func (b *customBalancer) Next(c echo.Context) *middleware.ProxyTarget {
	ctx := context.WithValue(c.Request().Context(), "FROM_BALANCER", "CUSTOM_BALANCER")
	c.SetRequest(c.Request().WithContext(ctx))
	return b.target
}

func TestModifyResponseUseContext(t *testing.T) {
	server := httptest.NewServer(
		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusOK)
			w.Write([]byte("OK"))
		}),
	)
	defer server.Close()

	targetURL, _ := url.Parse(server.URL)

	e := echo.New()
	e.Use(middleware.ProxyWithConfig(
		middleware.ProxyConfig{
			Balancer: &customBalancer{
				target: &middleware.ProxyTarget{
					Name: "tst",
					URL:  targetURL,
				},
			},
			RetryCount: 1,
			ModifyResponse: func(res *http.Response) error {
				val := res.Request.Context().Value("FROM_BALANCER")
				if valStr, ok := val.(string); ok {
					res.Header.Set("FROM_BALANCER", valStr)
				}
				return nil
			},
		},
	))

	req := httptest.NewRequest(http.MethodGet, "/", nil)
	rec := httptest.NewRecorder()

	e.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusOK, rec.Code)
	assert.Equal(t, "OK", rec.Body.String())
	assert.Equal(t, "CUSTOM_BALANCER", rec.Header().Get("FROM_BALANCER"))
}
```

Is it something like that?

Reply:
maybe

```go
				// This is needed for ProxyConfig.ModifyResponse and/or ProxyConfig.Transport to be able to process Request
				// that Balancer may have replaced with c.SetRequest.
				req = c.Request()
```

1. This comment says why and where if line is useful. 
2. do we need `req = req.WithContext()` at all we could just retake that pointer (pointing to new request instance) from echo.Context?

Reply:
also - please add at least one test

Reply:
Hi @aldas,

thanks for your review. I have adjusted the PR according to your comments.

> Is it something like that?

Yes, your test does exactly what I am trying to do.

> This comment says why and where if line is useful.

I have adopted the comment.

> do we need req = req.WithContext() at all we could just retake that pointer (pointing to new request instance) from echo.Context?

No, we don't need it. Your suggestion `req = c.Request()` is perfectly adequate.

> also - please add at least one test

I have adopted your test and fixed the context key lint error. Is that okay?